packages/typescript-estree/src/convert.ts

```diff
@@ -264,7 +271,7 @@ export class Converter {
     tsType: ts.TypeNode | undefined,
     parent?: ts.Node,
   ): TSESTree.BindingName {
-    const id = this.convertPattern(name);
+    const id = this.convertPattern(name) as TSESTree.BindingName;
```

Comment:
We could avoid these type casts if we can somehow add type-safety to `convertChild`, `convertPattern`, `convertType` but especially `converter` and make them generic.

Reply:
The logic is so messy and not type-safe because way back when it was written in pure JS - so it does a lot of things which you wouldn't usually do in TS.
Nobody's been motivated enough to rewrite it because the code works fine and is "safe enough"

Reply:
When I find some time, I'll try to take a look at it.

We can have the type casts for now I think.
It's just that while doing all of them, I really felt like I was doing the same thing over and over again and could be inferred somehow.